### ISSUE TITLE
Dans chaque page web, les messages de statut sont-ils correctement restitués par les technologies d'assistance ? (7.5)

### ISSUE THREAD

**Issue by tchak** [demarches-simplifiees/demarches-simplifiees.fr#6780]
# Dans chaque page web, les messages de statut sont-ils correctement restitués par les technologies d'assistance ?

## Messagerie

Par exemple :
- "Votre message a bien été envoyé à l’instructeur en charge de votre dossier."

Corrections :
- Pour les messages qui informent du résultat d'une action ou de l'état d'une application, ajouter le role="status".

### Création dossier

Par exemple :
- L'alerte qui indique que l'identité a bien été enregistrée
- Le message d'erreur " Le champ « Exemple d'une clé de routage : Choisir votre département de résidence » doit être rempli "
- Le message d'erreur lorsque le fichier ajouté n'est pas au bon format

Corrections :
- Pour les messages qui informent du résultat d'une action ou de l'état d'une application, ajouter le role="status"
- Pour les messages qui informent d'une alerte, ajouter le role="alert"

### Authentification

Par exemple :
- Le message qui indique "Courriel ou mot de passe incorrect. "

Corrections :
- Pour les messages qui présentent une suggestion ou avertissent d'une erreur, ajouter le role="alert".

**Comment by inseo** [comment]
# Préambule
Les zones live doivent impérativement être présentes au chargement de la page : elles ne peuvent pas être injectées à la volée en même temps que le message à restituer, faute de quoi ces derniers ne seront pas restitués.

L'attribut `aria-live="assertive"` (équivalent à un attribut `role="alert"`) est à utiliser pour les informations devant être restituées sans délai à l'utilisateur. Leur vocalisation est immédiate et interrompt l'utilisateur dans sa tâche.
L'attribut `aria-live="polite"` (équivalent à un attribut `role="status"`) est à utiliser pour les informations moins critiques et dont la vocalisation peut attendre que l'utilisateur soit inactif.   

Dans tous les cas l'utilisation de l'attribut `aria-live` doit s'accompagner de l'attribut [aria-atomic=”true”](https://developer.mozilla.org/fr/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#aria-atomic_) qui indique au lecteur d'écran qu'il doit présenter l'ensemble de la zone live, même si une partie seulement a été modifiée.

# État des lieux
Les message de status sont tous affichés dans le conteneur suivant : `<div aria-live="assertive" id="flash_messages">`.

Dans certains cas, ce conteneur reçoit comme contenu une nouvelle zone live : `<div role="status">`.   
Ce qui a pour effet de les imbriquer : 
```
<div id="flash_messages" aria-live="assertive">
  <div id="flash_message" class="center">
    <div class="alert alert-success" role="status">
      Vous êtes connecté(e) ! Vous pouvez à tout moment alterner entre vos différents profils : administrateur, instructeur, expert, usager.
    </div>
  </div>
</div>
```

# Correctif à apporter
La solution la plus simple à implémenter consiste certainement à prévoir deux zones live différentes : 
- la première pour les messages importants devant être restitués sans délai ;
- la seconde pour les messages informatifs pouvant attendre l'inactivité de l'utilisateur.

```
<div id="flash_messages">
  <div id="flash_message" class="center">
    <div role="alert"></div>
    <div role="status"></div>
  </div>
</div>
```

Suivant la nature du message, ce dernier sera injecté dans la zone live correspondante, en même temps que les classes lui conférant le rendu attendu (`class="alert alert-success"`, par exemple).

**Comment by inseo** [comment]
# Messages importants
Éléments à injecter dans la zone live `role="alert"`.

## users/dossiers_controller.rb
  * t('users.dossiers.no_access_html')
    => Vous n’avez pas accès à ce dossier.<br>Vérifiez que votre adresse email de connexion <b>%{email}</b> est bien celle utilisée pour remplir cette démarche.<br> Si ce n'est pas le cas, déconnectez-vous
  * t('users.dossiers.en_construction_submitted')
    => Les modifications ont déjà été déposées

## controllers/release_notes_controller.rb
  * t('release_notes.index.forbidden')
    => Vous n’êtes pas autorisé(e) à consulter la page des Nouveautés.

## instructeurs/batch_operations_controller.rb
  * Le traitement de masse n'a pas été lancé. Vérifiez que l'action demandée est possible pour les dossiers sélectionnés
  * Vous n’avez pas accès à cette démarche

## experts/avis_controller.rb
  * Vous n'êtes pas autorisé à acceder à la messagerie

## groupe_instructeurs_signature_concern.rb
  * Aucun fichier joint pour le tampon de l'attestation

## administrateurs/services_controller.rb
  * la démarche #{service_to_destroy.procedures.first.libelle} utilise encore le service #{service_to_destroy.nom}. Veuillez l'affecter à un autre service avant de pouvoir le supprimer
  * les démarches #{service_to_destroy.procedures.map(&:libelle).join(', ')} utilisent encore le service #{service.nom}. Veuillez les affecter à un autre service avant de pouvoir le supprimer

## users/sessions_controller.rb
  * Votre lien est invalide.
  * Votre lien est expiré, un nouveau vient de vous être envoyé.

## users/activate_controller.rb
  * Votre email est déjà vérifié
  * Ce lien n'est plus valable, un nouveau lien a été envoyé à l'adresse #{user.email}
  * Un problème est survenu, vous pouvez nous contacter sur #{Current.contact_email}

## manager/procedures_controller.rb
  * L’administrateur \"#{administrateur.email}\" est introuvable.
  * Importation impossible : veuillez importer un fichier CSV
  * Importation impossible : le poids du fichier est supérieur à #{number_to_human_size(CSV_MAX_SIZE)}
  * flash[:alert] = type_de_champ.errors.full_messages.join(', ')

## manager/gestionnaires_controller.rb
  * Impossible de supprimer ce gestionnaire car il est gestionnaire d'un groupe racine

## manager/administrateurs_controller.rb
  * t(reason, scope: "manager.administrateurs.delete")
    => L'administrateur ne peut pas être supprimé

## instructeurs/procedures_controller.rb
  * Vous n’avez pas accès à cette démarche
  * Votre affichage a dû être réinitialisé en raison du problème suivant : " + errors.full_messages.join(', ')

## instructeurs/groupe_instructeurs_controller.rb
  * L’instructeur « #{email} » est déjà dans le groupe.
  * L’adresse email « #{email} » n’est pas valide.
  * Suppression impossible : il doit y avoir au moins un instructeur dans le groupe
  * L’instructeur « #{instructeur.email} » n’est pas dans le groupe.
  * Vous n’avez pas le droit de gérer les instructeurs de cette démarche
  * Vous n’avez pas accès à cette démarche

## gestionnaires/groupe_gestionnaires_controller.rb
  * Impossible de supprimer ce groupe.

## manager/groupe_gestionnaires_controller.rb
  * t('activerecord.wrong_address')
    => %{emails} n’est pas une adresse email valide
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_gestionnaire.destroy_at_least_one')
    => Suppression impossible : il doit y avoir au moins un gestionnaire dans le groupe racine
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_gestionnaire.not_in_groupe_gestionnaire')
    => Le gestionnaire « %{email} » n’est pas dans le groupe gestionnaire.
  * t('activerecord.errors.duplicate_email')
    => {emails} est déjà gestionnaire de ce groupe

## gestionnaires/groupe_gestionnaire_gestionnaires_controller.rb
  * t('activerecord.wrong_address')
    => %{emails} n’est pas une adresse email valide
  * t('activerecord.errors.duplicate_email')
    => %{emails} est déjà gestionnaire de ce groupe
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_gestionnaire.destroy_at_least_one')
    => Suppression impossible : il doit y avoir au moins un gestionnaire dans le groupe racine
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_gestionnaire.not_in_groupe_gestionnaire')
    => Le gestionnaire « %{email} » n’est pas dans le groupe gestionnaire.

## gestionnaires/groupe_gestionnaire_administrateurs_controller.rb
  * t('activerecord.errors.administrateurs_already_in_groupe_gestionnaire')
    => Ajout impossible. Cet administrateur fait déjà partie d'un groupe.
  * t('activerecord.wrong_address')
    => %{emails} n’est pas une adresse email valide
  * t('activerecord.errors.duplicate_email')
    => %{emails} est déjà gestionnaire de ce groupe
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_administrateur.not_in_groupe_gestionnaire')
    => L'administrateur « %{email} » n’est pas dans le groupe gestionnaire.
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_administrateur.not_in_groupe_gestionnaire')
    => Le gestionnaire « %{email} » n’est pas dans le groupe gestionnaire.
  * t('groupe_gestionnaires.flash.alert.groupe_gestionnaire_administrateur.cannot_be_deleted')
    => L'administrateur « %{email} » ne peut pas être supprimé du groupe gestionnaire.

## administrateurs/groupe_instructeurs_controller.rb
  * Impossible de supprimer un groupe avec des dossiers. Il faut le réaffecter avant
  * Suppression impossible : il doit y avoir au moins un groupe instructeur sur chaque procédure
  * Suppression impossible : le groupe « #{@groupe_instructeur.label} » est le groupe par défaut.
  * Suppression impossible : il doit y avoir au moins un instructeur dans le groupe
  * Importation impossible : veuillez importer un fichier CSV
  * Importation impossible : le poids du fichier est supérieur à #{number_to_human_size(CSV_MAX_SIZE)}
  * Importation impossible, veuillez importer un csv suivant #{view_context.link_to('ce modèle', "/csv/import-instructeurs-test.csv")} pour une procédure sans routage ou #{view_context.link_to('celui-ci', "/csv/#{I18n.locale}/import-groupe-test.csv")} pour une procédure routée

## instructeurs/archives_controller.rb
## administrateurs/archives_controller.rb
  * Cette archive a déjà été générée.

## administrateurs/api_tokens_controller.rb
  * Vous ne pouvez pas supprimer les restrictions d'accès à l'API d'un jeton permanent.

## manager/administrateur_confirmations_controller.rb
  * Le lien que vous avez utilisé est invalide. Veuillez contacter la personne qui vous l'a envoyé.

**Comment by inseo** [comment]
# Reste à déterminer
## gestionnaires/groupe_gestionnaire_children_controller.rb
  * flash[:alert] = @child.errors.full_messages

## manager/procedures_controller.rb
  * flash[:alert] = type_de_champ.errors.full_messages.join(', ')

## manager/users_controller.rb
  * user.errors.full_messages.to_sentence


## administrateurs/ineligibilite_rules_controller.rb
  * flash[:alert] = draft_revision.errors.full_messages

## experts/avis_controller.rb
  * flash[:alert] = user.errors.full_messages

## instructeurs/contact_informations_controller.rb
  * flash[:alert] = @contact_information.errors.full_messages
  * flash[:alert] = @contact_information.errors.full_messages

## instructeurs/export_templates_controller.rb
  * flash[:alert] = @export_template.errors.full_messages
  * flash[:alert] = @export_template.errors.full_messages

## manager/confirmation_urls_controller.rb
  * flash[:alert] = alert

## manager/dossiers_controller.rb
  * flash[:alert] = transfer.errors.full_messages.join("<br>")

## shared_exemples_for_dossier.rb
  * expect(flash[:alert]).to be_present
